### PR TITLE
test-bot: don't consider optional deps for compiler selection

### PR DESCRIPTION
**Pull request by tdsmith**
Fixes test-bot mishandling of optional cross-tap dependencies. Without
this change, since formula.stable.deps includes the optional dependencies but
formula.recursive_dependencies does not, test-bot was trying to select a
compiler for an untapped formula, which raised an error.

Our suspect handling of optional dependencies was exposed by #43145 /
7184348e822d6745d8a103de104ff9626536bf1e.

### CODE REVIEW

**Comment by tdsmith** [comment]
casts summon on @xu-cheng

**Comment by tdsmith** [comment]
related to build failure in #43229

File: Library/Homebrew/cmd/test-bot.rb [hunk 1]
```diff
@@ -398,15 +398,15 @@ def formula(formula_name)
       if formula.stable
         return unless satisfied_requirements?(formula, :stable)
 
-        deps |= formula.stable.deps.to_a
-        reqs |= formula.stable.requirements.to_a
+        deps |= formula.stable.deps.to_a.select { |d| !d.optional? }
```
Comment (xu-cheng) [inline Library/Homebrew/cmd/test-bot.rb]: You can use `reject(&:optional?)`. :smile:

**Comment by xu-cheng** [comment]
This reminds me that we should do recursive_dependencies multiple times for different specs : stable devel and HEAD

**Comment by tdsmith** [comment]
thanks, will update tomorrow. i'm not sure i understand how the current code handles HEAD builds.

**Comment by mikemcquaid** [comment]
> i'm not sure i understand how the current code handles HEAD builds.

Probably: it doesn't!

**Comment by tdsmith** [comment]
I changed up the strategy a little, which should prevent the need for calling recursive_dependencies for different specs.

**Comment by mikemcquaid** [comment]
:+1:

**Comment by xu-cheng** [comment]
:+1: for this approach.

**Pull request closed without merging**